### PR TITLE
HostPID integration

### PR DESCRIPTION
**Pull request by pweil-**
Contains:

1. upstream patch for the HostPID support
1. v1beta3 carry for adding HostPID to v1beta3 api
1. the basic integration of the new admission controller
1. integration with old school SCCs - pending movement on the upstream PSP proposal

@liggitt @deads2k @pmorie @smarterclayton PTAL

@kargakis - FYI, hope this does mess up the rebase too bad if it goes in first, just didn't want to sit on it.   Let me know if you need assistance with anything that comes up related to HostPID or the DenyEscalatingExec admission controller.

### CODE REVIEW

**Comment by deads2k** [comment]
> @kargakis - FYI, hope this does mess up the rebase too bad if it goes in first, just didn't want to sit on it. Let me know if you need assistance with anything that comes up related to HostPID or the DenyEscalatingExec admission controller.

Is this important enough to get in ahead of the rebase?  We only did https://github.com/openshift/origin/pull/4703 because it prevented making connections to the server.

**Comment by pweil-** [comment]
> Is this important enough to get in ahead of the rebase? We only did #4703 because it prevented making connections to the server.

No, and I'm happy to rebase it after the rebase.  Just didn't want to sit on it while it was fresh in my head.  Reviews still welcome, not much will change in the Origin pieces.

**Comment by liggitt** [comment]
I wouldn't want this halfway in (exposed but not protected by SCC)

**Comment by pweil-** [comment]
> I wouldn't want this halfway in (exposed but not protected by SCC)

fine by me.  I need to work out the SCC piece anyway, I'll just lump it in here.  It will be fairly small since it's just a boolean field.

If the upstream commit doesn't move for PSP I'll have to integrate it into that PR as well anyways but it's already huge.

**Comment by liggitt** [comment]
Keep the SCC bit in a separate commit, but in this PR is fine

**Comment by pweil-** [comment]
```
[vagrant@openshiftdev pidtests]$ oc whoami
system:admin
[vagrant@openshiftdev pidtests]$ oc create -f pid.yaml 
pod "pid" created

[vagrant@openshiftdev pidtests]$ docker inspect 50c7 | grep Pid
        "PidMode": "host",
        "Pid": 31264,

[vagrant@openshiftdev pidtests]$ oc exec pid -- ps -ef | grep openshift
Error from server: Pod "Unknown" is forbidden: Cannot exec into or attach to a container using host pid
  
[vagrant@openshiftdev pidtests]$ docker exec 50c7 ps -ef | grep openshift
root     30776  1467  0 18:54 ?        00:00:00 sudo /data/src/github.com/openshift/origin/_output/local/go/bin/openshift --loglevel=4 start --latest-images=true --volume-dir=/tmp/origin.local.volumes
root     30777 30776 13 18:54 ?        00:00:24 /data/src/github.com/openshift/origin/_output/local/go/bin/openshift --loglevel=4 start --latest-images=true --volume-dir=/tmp/origin.local.volumes
1000     31335  9985  0 18:57 ?        00:00:00 grep --color=auto openshift

[vagrant@openshiftdev pidtests]$ oc whoami
test
[vagrant@openshiftdev pidtests]$ oc create -f pid.yaml 
Error from server: error when creating "pid.yaml": Pod "pid" is forbidden: unable to validate against any security context constraint: [provider restricted: .spec.containers[0].securityContext.hostPID: invalid value 'true', Details: Host PID is not allowed to be used]
```

**Comment by pmorie** [comment]
`provider restricted: .spec.containers[0].securityContext.hostPID`

@pweil- Why is this on a container's security context?

**Comment by pweil-** [comment]
> @pweil- Why is this on a container's security context?

The admission controller is an error from the provider lumping them all together with:
```
append(errs, provider.ValidateSecurityContext(pod, &c).Prefix(fmt.Sprintf("spec.containers[%d].securityContext", i))...)
```  
It would have to inspect the error to determine whats in the SC vs. what isn't.  Or perhaps it shouldn't even try and just return whatever it gets.

**Comment by pweil-** [comment]
Yes, I like the idea of letting the admission controller not worry about it after giving it a quick test.  I'll take that as a follow up since it is related more to the overall SCC/PSP provider.  I'll want to make all the errors in there a bit more specific as to where the field comes from.

**Comment by smarterclayton** [comment]
Can we land your downstream PSP and mark it experimental?  We can always
drop it prior to 1.1

On Sep 17, 2015, at 2:29 PM, Paul Weil <notifications@github.com> wrote:

I wouldn't want this halfway in (exposed but not protected by SCC)

fine by me. I need to work out the SCC piece anyway, I'll just lump it in
here. It will be fairly small since it's just a boolean field.

If the upstream commit doesn't move for PSP I'll have to integrate it into
that PR as well anyways but it's already huge.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4712#issuecomment-141177945>.

**Comment by pweil-** [comment]
> Can we land your downstream PSP and mark it experimental?  We can always
> drop it prior to 1.1

I guess it depends on what you mean by marking it experimental.  SCC is not experimental now so that seems like a regression.

I'm mostly worried about maintaining backwards compatibility if we move SCC -> PSP downstream.  In the current PR I'm doing a conversion from SCC->PSP in a proxy storage impl.  If the PSP api upstream changes then we'll need PSP (OS) -> PSP (Kube) as well if we want 100% backwards compat.  

The other option is to continue to carry SCC until PSP lands upstream.  We'll still need the SCC->PSP conversion anyway but it would eliminate the possibility of a second conversion.  I really have my fingers crossed that upstream will accept the current proposal in their exp api.

**Comment by smarterclayton** [comment]
SCC would be converted and not experimental.  We have to carry SCC for a
while now, so your code is going to be required for a long time.  I just
want to ensure that we have a path forward before the end of this sprint.

On Thu, Sep 17, 2015 at 3:55 PM, Paul Weil <notifications@github.com> wrote:

> Can we land your downstream PSP and mark it experimental? We can always
> drop it prior to 1.1
>
> I guess it depends on what you mean by marking it experimental. SCC is not
> experimental now so that seems like a regression.
>
> I'm mostly worried about maintaining backwards compatibility if we move
> SCC -> PSP downstream. In the current PR I'm doing a conversion from
> SCC->PSP in a proxy storage impl. If the PSP api upstream changes then
> we'll need PSP (OS) -> PSP (Kube) as well if we want 100% backwards compat.
>
> The other option is to continue to carry SCC until PSP lands upstream.
> We'll still need the SCC->PSP conversion anyway but it would eliminate the
> possibility of a second conversion. I really have my fingers crossed that
> upstream will accept the current proposal in their exp api.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4712#issuecomment-141207369>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by smarterclayton** [comment]
The experimental PSP in origin we could nuke without warning if it was so
marked

On Thu, Sep 17, 2015 at 5:09 PM, Clayton Coleman <ccoleman@redhat.com>
wrote:

> SCC would be converted and not experimental.  We have to carry SCC for a
> while now, so your code is going to be required for a long time.  I just
> want to ensure that we have a path forward before the end of this sprint.
>
> On Thu, Sep 17, 2015 at 3:55 PM, Paul Weil <notifications@github.com>
> wrote:
>
>> Can we land your downstream PSP and mark it experimental? We can always
>> drop it prior to 1.1
>>
>> I guess it depends on what you mean by marking it experimental. SCC is
>> not experimental now so that seems like a regression.
>>
>> I'm mostly worried about maintaining backwards compatibility if we move
>> SCC -> PSP downstream. In the current PR I'm doing a conversion from
>> SCC->PSP in a proxy storage impl. If the PSP api upstream changes then
>> we'll need PSP (OS) -> PSP (Kube) as well if we want 100% backwards compat.
>>
>> The other option is to continue to carry SCC until PSP lands upstream.
>> We'll still need the SCC->PSP conversion anyway but it would eliminate the
>> possibility of a second conversion. I really have my fingers crossed that
>> upstream will accept the current proposal in their exp api.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/openshift/origin/pull/4712#issuecomment-141207369>.
>>
>
>
>
> --
> Clayton Coleman | Lead Engineer, OpenShift
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by pweil-** [comment]
Ok, I think I'm on the same page.  I think what I have in the other PR works then.  I'll try to break it up to reduce the eye bleeding reviews into PRs for

1.  straight conversion from kube SCC -> OS SCC
2.  OS PSP type with conversion to SCC (all components still working with SCC but PSP api is introduced)
3.  Change of the admission/bootstrap/etc to use the OS PSP type

**Comment by pweil-** [comment]
This is cleaned up and rebased with support for host ipc and host pid.  I would like to get this reviewed and merged regardless of the state of an experimental PSP object.  This provides functionality while the PSP object just gets us closer to the proposal upstream and isn't crucial.

File: Godeps/_workspace/src/k8s.io/kubernetes/cmd/kube-apiserver/app/plugins.go [hunk 1]
```diff
@@ -26,7 +26,7 @@ import (
 	// Admission policies
 	_ "k8s.io/kubernetes/plugin/pkg/admission/admit"
 	_ "k8s.io/kubernetes/plugin/pkg/admission/deny"
-	_ "k8s.io/kubernetes/plugin/pkg/admission/exec/denyprivileged"
```
Comment (pmorie) [inline Godeps/_workspace/src/k8s.io/kubernetes/cmd/kube-apiserver/app/plugins.go]: Why is this being removed?

File: Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubelet/util.go [hunk 2]
```diff
@@ -73,3 +93,31 @@ func allowHostNetwork(pod *api.Pod) (bool, error) {
 	}
 	return false, nil
 }
+
+// Determined whether the specified pod is allowed to use host networking
```
Comment (pmorie) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubelet/util.go]: Typo: determine

File: Godeps/_workspace/src/k8s.io/kubernetes/cmd/kube-apiserver/app/plugins.go (hunk 1 above)
Comment (pweil-) [inline Godeps/_workspace/src/k8s.io/kubernetes/cmd/kube-apiserver/app/plugins.go]: deprecated upstream in favor of `DenyEscalatingExec` which is covered by the line below (which also registers itself under the old plugin name)

File: Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubelet/util.go [hunk 3]
```diff
@@ -73,3 +93,31 @@ func allowHostNetwork(pod *api.Pod) (bool, error) {
...
+	}
+	for _, source := range capabilities.Get().PrivilegedSources.HostPIDSources {
+		if source == podSource {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+// Determined whether the specified pod is allowed to use host ipc
```
Comment (pmorie) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubelet/util.go]: Same typo.  We probably don't need to worry about fixing it unless there's something else substantial to fix about this.

**Comment by pmorie** [comment]
LGTM

**Comment by pmorie** [comment]
[test]

**Comment by deads2k** [comment]
jenkins issue re[test]

**Comment by pweil-** [comment]
bump for reviews.  I'd like to get this in since the PSP types is still under heavy review.  Travis seemed to be failing on an unrelated command in test-cmd.  Rebased to see if that fixes.

[test]

**Comment by smarterclayton** [comment]
API changes arenapproved

**Comment by pweil-** [comment]
@pmorie @deads2k - please take another pass? The commits are in pretty digestible chunks

**Comment by pweil-** [comment]
tagging some extra reviewers @soltysh @miminar

**Comment by smarterclayton** [comment]
This is LGTM [merge]

**Comment by pmorie** [comment]
Looks like jenkins flaked and this LGTM

[merge]

**Comment by pweil-** [comment]
This appears to have flaked again.  I am going to rebase it and re-run the tests locally, I have been unable to reproduce the failures.

**Comment by pweil-** [comment]
@pmorie another unrelated k8s flake, seen this one before:

```
2015/10/13 05:00:43 http: panic serving 127.0.0.1:60159: bad read
goroutine 52 [running]:
net/http.func·011()
	/usr/local/go/src/net/http/server.go:1130 +0xcc
github.com/openshift/origin/Godeps/_workspace/src/k8s.io/kubernetes/pkg/util/wsstream.(*errorReader).Read(0xc2080882c0, 0xc20813c800, 0x800, 0x800, 0xc208086ae0, 0x0, 0x0)
	/home/travis/gopath/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/Godeps/_workspace/src/k8s.io/kubernetes/pkg/util/wsstream/stream_test.go:151 +0x129
github.com/openshift/origin/Godeps/_workspace/src/k8s.io/kubernetes/pkg/util/wsstream.messageCopy(0xc208084090, 0x2aba79740200, 0xc2080882c0, 0xc208080000, 0x0, 0x0, 0x0)
	github.com/openshift/origin/Godeps/_workspace/src/k8s.io/kubernetes/pkg/util/wsstream/_test/_obj_test/stream.go:105 +0x2aa
github.com/openshift/origin/Godeps/_workspace/src/k8s.io/kubernetes/pkg/util/wsstream.(*Reader).handle(0xc208044ff0, 0xc208084090)
	github.com/openshift/origin/Godeps/_workspace/src/k8s.io/kubernetes/pkg/util/wsstream/_test/_obj_test/stream.go:79 +0x2c6
github.com/openshift/origin/Godeps/_workspace/src/k8s.io/kubernetes/pkg/util/wsstream.func·009(0xc208084090)
	/home/travis/gopath/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/Godeps/_workspace/src/k8s.io/kubernetes/pkg/util/wsstream/stream_test.go:171 +0x1aa
golang.org/x/net/websocket.Server.serveWebSocket(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xa2fcc8, ...)
	/home/travis/gopath/src/github.com/openshift/origin/Godeps/_workspace/src/golang.org/x/net/websocket/server.go:89 +0x2db
golang.org/x/net/websocket.Handler.ServeHTTP(0xc2080866a0, 0x2aba7973fd28, 0xc208075040, 0xc20801e680)
	/home/travis/gopath/src/github.com/openshift/origin/Godeps/_workspace/src/golang.org/x/net/websocket/server.go:112 +0x9b
net/http/httptest.(*waitGroupHandler).ServeHTTP(0xc208086820, 0x2aba7973fd28, 0xc208075040, 0xc20801e680)
	/usr/local/go/src/net/http/httptest/server.go:200 +0xf8
net/http.serverHandler.ServeHTTP(0xc2080306c0, 0x2aba7973fd28, 0xc208075040, 0xc20801e680)
	/usr/local/go/src/net/http/server.go:1703 +0x1f7
net/http.(*conn).serve(0xc208074e60)
	/usr/local/go/src/net/http/server.go:1204 +0x1088
created by net/http.(*Server).Serve
	/usr/local/go/src/net/http/server.go:1751 +0x3cd
--- FAIL: TestStreamClosedDuringRead-2 (0.24s)
	stream_test.go:131: unexpected server read: []
FAIL
coverage: 70.5% of statements
FAIL	github.com/openshift/origin/Godeps/_workspace/src/k8s.io/kubernetes/pkg/util/wsstream	0.729s
```

**Comment by pmorie** [comment]
Light is green, trap is clean, I ain't afraid of no :ghost: 

[merge]

**Pull request merged**